rule_test.go

```diff
@@ -92,7 +92,7 @@ func TestContentFormatPattern(t *testing.T) {
 			input: &Content{
 				Pattern: []byte(`C:\\WINDOWS\\system32\\`),
 			},
-			want: `C|3A|\\WINDOWS\\system32\\`,
+			want: `C|3A 5C 5C|WINDOWS|5C 5C|system32|5C 5C|`,
```

Comment:
I don't think this is correct.  If the input from the Suricata rule text is: `C:\\WINDOWS\\system32\\` we really want the output to be `C|3A 5C|WINDOWS|5C|system32|5C` or `C:\\WINDOWS\\system32\\`

outputting |5C 5C| would equate to a rule that is `C|3A|\\\\WINDOWS\\\\system32\\\\`.

In order to address this issue, I think the parsing logic in parseContent probably needs to be updated to handle this conversion. Probably this should consider any other characters that might be escaped in a rule as well.

If you'd like to continue on this, I'd suggest adding a test to `TestParseContent()` in `parser_test.go` that looks like this:
```
{
	name:  "contains escaped slash",
	input: `A\\B`,
	want:  []byte(`A\B`),
},
```

Reply:
You're correct. I added handling of escaped characters (`\`, `;`, `"`, `:`) to `parseContent`.

Based on this note in section of [docs](https://suricata.readthedocs.io/en/latest/rules/intro.html#rule-options) escaped characters should be `;`, `"`, `\` but based on the [suricata source code](https://github.com/OISF/suricata/blob/master/src/detect-content.c#L150), `:` should be there as well - this can be also seen in tests in the same file ([1](https://github.com/OISF/suricata/blob/master/src/detect-content.c#L763), [2](https://github.com/OISF/suricata/blob/master/src/detect-content.c#L794), [3](https://github.com/OISF/suricata/blob/master/src/detect-content.c#L825) and [4](https://github.com/OISF/suricata/blob/master/src/detect-content.c#L856)).

Do you think gonids should also give some [escape error](https://github.com/OISF/suricata/blob/master/src/detect-content.c#L159) when appropriate? I tried few contents in suricata and I think even suricata doesn't handle these errors correctly - e.g. your test `content with backslash at end` of `TestParseRule` is probably invalid rule (`content:"foo\"`), but suricata loads the rule without an error, however for `content:"foo\bar"` it gives an error.

Reply:
This looks great, thanks!
Yes I think we should raise an error if the rule is incorrect. One of the hopes is that this could be leveraged in developing a rule linter of sorts to flag incorrect syntax as an author is writing a rule. Raising an error is very much in line with that use case and others.

Reply:
Hi, have you had time to check if the solution from last weeks' commit is ok?

Reply:
Sorry, work and personal has been crazy lately. Apologies for the delay. This looks great, thanks!